lib/framework/procbot.ts

```diff
@@ -108,58 +118,52 @@ export class ProcBot {
 
 	/**
 	 * Add a new type of ServiceListener to the client.
-	 * Should the ServiceListener already exist on the client, this will do nothing.
+	 * Should the ServiceListener already exist on the client with the handle/name, this will do nothing.
+	 *
 	 * @param name   The name of the ServiceListener to add.
-	 * @param data?  Any relevant data required to construct the ServiceListener.
+	 * @param data?  A ServiceBase extended constructor, should data be passed to a Service constructor.
 	 * @return       The constructed ServiceListener or void should it already exist or fail.
 	 */
 	protected addServiceListener(name: string, data?: any): ServiceListener | void {
-		const service = this.getService(name);
-		let listener;
-
-		if (service && !_.find(this.listeners, [ 'serviceName', name ])) {
-			listener = service.createServiceListener(data);
-			this.listeners.push(listener);
-		}
-
-		return listener;
+		return <ServiceListener>this.addService('listener', name, data);
```

Comment:
@CameronDiver demo with enums [here](https://www.typescriptlang.org/play/#src=enum%20ServiceType%20%7B%0A%20%20%20%20Listener%20%3D%200%2C%0A%20%20%20%20Emitter%20%3D%201%0A%7D%20%0Atype%20ServiceEmitter%20%3D%20%7B%7D%0Atype%20ServiceListener%20%3D%20%7B%7D%0Atype%20ServiceConstructor%20%3D%20%7B%7D%0A%0Afunction%20addService(type%3A%20ServiceType.Listener%2C%20name%3A%20string)%3A%20ServiceListener%20%7C%20void%0Afunction%20addService(type%3A%20ServiceType.Emitter%2C%20name%3A%20string)%3A%20ServiceEmitter%20%7C%20void%0Afunction%20addService(type%3A%20ServiceType%2C%20name%3A%20string)%3A%20ServiceListener%20%7C%20ServiceEmitter%20%7C%20void%0A%7B%0A%0A%7D%0A%0A%2F%2F%20Mouseover%20vars%20to%20see%20the%20inferred%20types%0Avar%20a%20%3D%20addService(ServiceType.Listener%2C%20'a')%3B%0Avar%20e%20%3D%20addService(ServiceType.Emitter%2C%20'b')%3B).

You can do the same thing with string literals or anything else TS can match at a type level. You just write a series of function signatures together, it uses the first one it finds that matches in the order they're defined, and the last one has to have a broad enough signature to encompass all the possible previous types, and is what the actual function body has to match.